### PR TITLE
Fix error codes from various RPCs

### PR DESCRIPTION
**Pull request by jnewbery**
(Completes #9707 and #9842)

I've changed the qa test code to properly assert error codes and messages for failed RPC calls. That's revealed a few RPCs which were returning incorrect or misleading error codes. This PR fixes all of those cases.

It also adds commenting around the definition of the JSON RPC error constants to warn against using RPC_INVALID_REQUEST and RPC_METHOD_NOT_FOUND for application-layer errors.

### CODE REVIEW

File: qa/rpc-tests/bumpfee.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def test_segwit_bumpfee_succeeds(rbf_node, dest_address):
 def test_nonrbf_bumpfee_fails(peer_node, dest_address):
     # cannot replace a non RBF transaction (from node which did not enable RBF)
     not_rbfid = create_fund_sign_send(peer_node, {dest_address: 0.00090000})
-    assert_raises_message(JSONRPCException, "not BIP 125 replaceable", peer_node.bumpfee, not_rbfid)
+    assert_raises_jsonrpc(-4, "not BIP 125 replaceable", peer_node.bumpfee, not_rbfid)
```
Comment (ryanofsky) [inline qa/rpc-tests/bumpfee.py]: Is there a plan to switch from integer literals to constants in a later PR? Seems like it would be nice to say `RPC_WALLET_ERROR` instead of `-4`.

Comment (laanwj) [inline qa/rpc-tests/bumpfee.py]: Yes that would be nice, but is orthogonal to this change, so if you want to do that it'd indeed be a different PR.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2753,33 +2753,33 @@ UniValue bumpfee(const JSONRPCRequest& request)
     CWalletTx& wtx = pwalletMain->mapWallet[hash];
 
     if (pwalletMain->HasWalletSpend(hash)) {
-        throw JSONRPCError(RPC_MISC_ERROR, "Transaction has descendants in the wallet");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Transaction has descendants in the wallet");
```
Comment (ryanofsky) [inline src/wallet/rpcwallet.cpp]: In cases where bumpfee is called erroneously with a transaction that shouldn't be bumped, invalid parameter seems more apt than wallet error.

Probably the most important thing you want to know when you get an error message from an API is whether the error is caused by something that you control and need to fix, or whether it's caused by something out of your control (like an implementation bug, or corrupt data, or a temporary failure). RPC_INVALID_PARAMETER clearly tells you that you need to fix something, while RPC_WALLET_ERROR is vague but suggests that something's wrong with the wallet.

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -338,11 +338,11 @@ UniValue removeprunedfunds(const JSONRPCRequest& request)
     vector<uint256> vHashOut;
 
     if(pwalletMain->ZapSelectTx(vHash, vHashOut) != DB_LOAD_OK) {
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Could not properly delete the transaction.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Could not properly delete the transaction.");
     }
 
     if(vHashOut.empty()) {
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Transaction does not exist in wallet.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Transaction does not exist in wallet.");
```
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: Maybe invalid parameter here as well.

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -338,11 +338,11 @@ UniValue removeprunedfunds(const JSONRPCRequest& request)
     vector<uint256> vHashOut;
 
     if(pwalletMain->ZapSelectTx(vHash, vHashOut) != DB_LOAD_OK) {
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Could not properly delete the transaction.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Could not properly delete the transaction.");
```
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: ZapSelect returns a DBErrors enum value which seems like it could be mapped to an RPC error (would also be nice to include in the error string).

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -829,7 +834,7 @@ UniValue pruneblockchain(const JSONRPCRequest& request)
             + HelpExampleRpc("pruneblockchain", "1000"));
 
     if (!fPruneMode)
-        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Cannot prune blocks because node is not in prune mode.");
+        throw JSONRPCError(RPC_MISC_ERROR, "Cannot prune blocks because node is not in prune mode.");
```
Comment (ryanofsky) [inline src/rpc/blockchain.cpp]: Some of these MISC_ERROR cases make it seem like we could benefit from an new RPC_INVALID_STATE value. (Blockchain too short is another case below.)

**Review by ryanofsky — COMMENTED** [review]
utACK b34faf156d49705faa335641cfbfc0efd60bb781, definitely an improvement.

File: qa/rpc-tests/bumpfee.py (hunk 1 above)
Comment (jnewbery) [inline qa/rpc-tests/bumpfee.py]: yep, I'll put that in the todo list. I'm working on a primitives.py for bitcoin messages, datastructures and constants. Seems like a good place for the RPC error constants to live.

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (jnewbery) [inline src/wallet/rpcwallet.cpp]: > Probably the most important thing you want to know when you get an error message from an API is whether the error is caused by something that you control and need to fix, or whether it's caused by something out of your control

Nicely put. I'll update this to RPC_INVALID_PARAMETER

File: src/wallet/rpcdump.cpp (hunk 4 above)
Comment (jnewbery) [inline src/wallet/rpcdump.cpp]: I agree we could do better, but let's leave that for a future PR.

I think this is strictly an improvement because currently the API is returning RPC_INTERNAL_ERROR even in cases where there's no internal error.

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (jnewbery) [inline src/wallet/rpcdump.cpp]: ACK

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (jnewbery) [inline src/rpc/blockchain.cpp]: Sounds sensible, but again let's leave that for a future PR.

Comment (laanwj) [inline src/rpc/blockchain.cpp]: Yes, there is no reason why we couldn't add new status values. There's no need to make everything fit in the arbitrary bag of codes that exists now, which has been there unchanged for a long while.

**Comment by jnewbery** [comment]
Squashed @ryanofsky's nits and rebased.
Old tip: https://github.com/bitcoin/bitcoin/pull/9853/commits/5158a1e2a2587f9eba04d0f2a81744a9ecbe766d / https://github.com/jnewbery/bitcoin/tree/pr9853.1
New tip: https://github.com/bitcoin/bitcoin/pull/9853/commits/3ff485e32f7a1d068e799d323d3531ede22873f1 / https://github.com/jnewbery/bitcoin/tree/pr9853.2

**Comment by laanwj** [comment]
utACK 3ff485e

**Comment by MarcoFalke** [comment]
Concept ACK.

Style Nit: In blockchain.cpp you are missing braces around the
multiline if block

Also, I was wondering if this requires a mention in the release notes...

**Comment by laanwj** [comment]
> Also, I was wondering if this requires a mention in the release notes...

It does.

**Comment by ryanofsky** [comment]
utACK 3ff485e32f7a1d068e799d323d3531ede22873f1

Confirmed only minor changes since previous ACK.

**Comment by jnewbery** [comment]
Release notes updated. Otherwise no change.
